### PR TITLE
[KYUUBI #3371] [FEATURE] [AUTHZ] Support checking access privileges in single call and throws all disallowed in exception

### LINKED ISSUE #3371: [FEATURE] [AUTHZ] Support checking access privileges in single call and throws all disallowed in exception
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Currently Authz plugin checks all required access privileges one by one and fast fails with one single disallowed access. For ad-hoc query or data prepration senario, it's often neccessary to know all the unsatisfied access privileges for further processes like applying access, authorization and etc.

Here to suggest,
1. introduce an ranger plugin config for enabling full access privileges in Authz
2. Authz checks all access privileges and throw exception with disallowed messages, if the no.1 config enabled

The implementation also relies on #3361 for batch verifying for better performance.
### Motivation

_No response_

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by bowenliang123**
### _Why are the changes needed?_

to close #3371 

Adding an ranger plugin config for enabling full access privileges,  Authz checks all access privileges and throw exception with message of unsatisfied  privileges on different resources in natural order in execution plan.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by bowenliang123** [comment]
I would like to move a new session config "kyuubi.authz.enable.full.access.check" from spark local property setting to KyuubiConf. How to make it possible?

**Comment by bowenliang123** [comment]
cc @yaooqinn 
Code changes and added config doc are ready for review. Please have a check.

File: docs/security/authorization/spark/install.md [hunk 1]
```diff
@@ -92,6 +92,16 @@ Macros are now supported for using user/group/tag in row filter expressions, int
     </property>    
 ```
 
+##### Checking all required Access privileges
```
Comment (yaooqinn) [inline docs/security/authorization/spark/install.md:95]: the naming and meaning of this configuration are ambiguous.

No matter whether it is enabled or disabled, we need to check all required privileges.

**Comment by yaooqinn** [comment]
let us simply do

```scala
val requestArrays = requests.map { request =>
      val resource = request.getResource.asInstanceOf[AccessResource]
      resource.objectType match {
        case ObjectType.COLUMN if resource.getColumns.nonEmpty =>
          resource.getColumns.map { col =>
            val cr = AccessResource(COLUMN, resource.getDatabase, resource.getTable, col)
            AccessRequest(cr, ugi, opType, request.accessType).asInstanceOf[RangerAccessRequest]
          }
        case _ => Seq(request)
      }
    }

    if (SparkRangerAdminPlugin.authorizeInSingleRpc) {
      verify(requestArrays.flatten.asJava, auditHandler)
    } else {
      requestArrays.foreach(requests => verify(requests.asJava, auditHandler))
    }
```

```scala
... SparkRangerAdminPlugin.scala
  /**
   * For a Spark SQL query, it may contain 0 or more privilege objects to verify, e.g. a typical
   * JOIN operator may have two tables and their columns to verify.
   *
   * This configuration controls whether to verify the privilege objects in a single RPC call or
   * to verify them one by one.
   */
  val authorizeInSingleRpc: Boolean =
    getConfig.getBoolean(s"ranger.plugin.${getServiceType}.authorize.in.single.rpc", false)
}

````

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 2]
```diff
@@ -76,4 +94,54 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql") {
     val digits = s"regexp_replace($lower, '[0-9]', 'n'$pos)"
     digits
   }
+
+  @throws[AccessControlException]
+  def verify(
+      requests: util.List[RangerAccessRequest],
+      auditHandler: SparkRangerAuditHandler,
+      enabledFullAccessViolationMsg: Boolean): Unit = {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:102]: let's revert this change in this PR

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 3]
```diff
@@ -17,12 +17,30 @@
...
+import org.apache.kyuubi.plugin.spark.authz.AccessControlException
 import org.apache.kyuubi.plugin.spark.authz.util.AuthZUtils._
 
 object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql") {
 
+  /**
+   * For a Spark SQL query, it may contain 0 or more privilege objects to verify, e.g. a typical
+   * JOIN operator may have two tables and their columns to verify.
+   *
+   * This configuration controls whether to verify the privilege objects in a single RPC call or
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:38]: nit:a single ~RPC~ call

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 4]
```diff
@@ -76,4 +94,38 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql") {
     val digits = s"regexp_replace($lower, '[0-9]', 'n'$pos)"
     digits
   }
+
+  @throws[AccessControlException]
+  def verify(
+      requests: util.List[RangerAccessRequest],
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:100]: import java.util.List

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:100]: When with imported ` java.util.List`, a format warning is shown by IDEA. Does it matter?
image

If autofix by IDEA with prefix, falling back to `util.List`.

Comment (pan3793) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:100]: `util.List` is good, to distinguish w/ scala List

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:106]: OK. let's keep it with util prefix.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 3 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: fixed.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 2 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: ok, applied.

File: docs/security/authorization/spark/install.md (hunk 1 above)
Comment (bowenliang123) [reply docs/security/authorization/spark/install.md]: Changed to `Showing all disallowed privileges`.

File: docs/security/authorization/spark/install.md [hunk 5]
```diff
@@ -92,6 +92,16 @@ Macros are now supported for using user/group/tag in row filter expressions, int
     </property>    
 ```
 
+##### Showing all disallowed privileges
+By default, Authz plugin checks required privileges one by one and throw the first unsatisfied privilege in exception. By setting `ranger.plugin.spark.authorize.in.single.call` to `true`, Authz plugin executes access checks in single call and throws all disallowed privileges in exception message.
+```xml
+    <property>
+        <name>ranger.plugin.spark.authorize.in.single.call</name>
+        <value>true</value>
```
[the remark this answers is not in the document]
Comment (bowenliang123) [reply docs/security/authorization/spark/install.md:100]: Yes, it should be expected as before with default false value.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 6]
```diff
@@ -17,12 +17,27 @@
...
 
+  /**
+   * For a Spark SQL query, it may contain 0 or more privilege objects to verify, e.g. a typical
+   * JOIN operator may have two tables and their columns to verify.
+   *
+   * This configuration controls whether to verify the privilege objects in single call or
+   * to verify them one by one.
+   */
+  def authorizeInSingleCall: Boolean =
+    getConfig.getBoolean(s"ranger.plugin.${getServiceType}.authorize.in.single.call", false)
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:39]: this fails ranger 0.6

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 7]
```diff
@@ -76,4 +91,40 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql") {
...
+    if (requests.nonEmpty) {
+      val results = SparkRangerAdminPlugin.isAccessAllowed(requests.asJava, auditHandler)
+      if (results != null) {
+        val indices = results.asScala.zipWithIndex.filter { case (result, idx) =>
+          result != null && !result.getIsAllowed
+        }.map(_._2)
+        if (indices.nonEmpty) {
+          val user = requests.head.getUser
+          val accessTypeToResource =
+            indices.foldLeft(mutable.LinkedHashMap.empty[String, mutable.Set[String]])((m, idx) => {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:113]: nit: import

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: ok

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 6 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: It should be alright now with added `RangerConfigProvider`. Successful built with Ranger 0.6 locally, with failed test in `show functions` case which is not changed in this PR. Tests and builds are good with higher ranger versions.

File: docs/security/authorization/spark/install.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bowenliang123) [reply docs/security/authorization/spark/install.md:100]: True value here aims to show samples for how to enable this feature.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master
